### ISSUE TITLE
Invalid warnings and memory error on empty Doxygen strings

### ISSUE THREAD

**Issue by geographika** [swig/swig#1949]
The following is a comment in RST on a C struct. The word immutable should be in bold when processed in Sphinx so is wrapped in double asterisks. The \ is required to escape to ensure correct output. 

```c++
    int width; ///< \**immutable** image width in pixels
```

When run with the following command parameters:

`swig -python -py3 -doxygen -shadow -o test_wrap.c test.i`

the following line throws an unhandled memory exception when debugging in Visual Studio as the string is empty:

https://github.com/swig/swig/blob/ac8de714af1d7fbd755597b0f51792212f552ef7/Source/Doxygen/doxyparser.cxx#L1089

`Unhandled exception at 0x00007FF84489B7EC (ucrtbased.dll) in swig.exe: An invalid parameter was passed to a function that considers invalid parameters fatal.`

This is thrown in the `xsstring` module with the error `cannot dereference string iterator because it is out of range (e.g. an end iterator)`

The empty string is created here (`endOfWordPos` and `pos` are both 2):

https://github.com/swig/swig/blob/ac8de714af1d7fbd755597b0f51792212f552ef7/Source/Doxygen/doxyparser.cxx#L1181

In addition to the memory error, erroneous errors are output by SWIG to the command line:

```
..\swiginc\class.i(164) : Warning 560: Unknown Doxygen command: .
..\swiginc\class.i(179) : Warning 560: Unknown Doxygen command: .
..\swiginc\class.i(186) : Warning 560: Unknown Doxygen command: .
..\swiginc\class.i(191) : Warning 560: Unknown Doxygen command: .
```

**Comment by vadz** [comment]
I'm not sure why is backslash required, but in any case this shouldn't crash, of course. But I can't reproduce the problem with just
```diff
diff --git a/Examples/test-suite/doxygen_parsing.i b/Examples/test-suite/doxygen_parsing.i
index 40f37a4c2..864fdb884 100644
--- a/Examples/test-suite/doxygen_parsing.i
+++ b/Examples/test-suite/doxygen_parsing.i
@@ -2,6 +2,8 @@

 %inline %{

+int width; ///< \**immutable** image width in pixels
+
 /**
  * The class comment
  */
```
Could you please make a minimal patch to this test case showing it? TIA!

More generally, we really need to run at least some SWIG CI builds using ASAN...

**Comment by geographika** [comment]
@vadz thanks for the feedback. I'll update the test case in the pull request. 
The / was required or one of the * got dropped in the output Python docstring. 

Note when running swig.exe with the same input there is no crash (but still erroneous logging output). I believe when running through the Visual Studio debugger it is much stricter with memory leaks.

**Comment by vadz** [comment]
I think the fact that `*` got dropped could be a bug on its own, or is there any reason why this would be expected/correct behaviour?

As for the test, I don't see any error output with it neither. And I'm running it under ASAN which should normally detect any invalid memory accesses (although perhaps it doesn't detect dereferencing an invalid iterator if it still references readable memory, I'll have to double check this).

**Comment by geographika** [comment]
I'll check the `*` again now I have SWIG running in the debugger. 

Are you running on Windows? The error is thrown in the xstring module [here](https://github.com/microsoft/STL/blob/834baa6acacb8325367e255b0728c582c7e443ff/stl/inc/xstring#L1859), so maybe MSVC specific?

That link also shows that it will only ever be thrown at a certain debug level: `#if _ITERATOR_DEBUG_LEVEL >= 1` so likely not a major cause for concern.

**Comment by vadz** [comment]
No, I didn't test it with MSVC yet, but the bug should be also visible under Unix, dereferencing an invalid iterator is a problem anywhere, even if there is no assert/debugging check for it.

Just to be clear, do you actually see this error with MSVC with just the patch above?

**Comment by geographika** [comment]
Strangely no, but I can see it with the following:

```diff
--- a/Examples/test-suite/doxygen_parsing.i
+++ b/Examples/test-suite/doxygen_parsing.i
@@ -29,6 +29,7 @@ enum SomeEnum
  */
 struct SomeStruct
 {
+    int width; ///< \**immutable** image width in pixels
 };
 ```

**Comment by vadz** [comment]
OK, thanks, I can reproduce it with this change. ASAN indeed doesn't detect this because the iterator, which is just a pointer here, is actually valid and points to the trailing NUL, but it's still undefined behaviour to dereference it, so we should avoid doing it.

I think #1952 might be a slightly better fix as there are already similar tests for `!cmd.empty()` in the existing code.

But the actual problem is that this backslash shouldn't be needed here in the first place, it doesn't escape anything and the fact that it "works" is completely accidental. The asterisks are getting stripped because the code always removes them in the beginning of (each line of) a comment, to deal with the semi-standard
```cpp
/**
 * Whatever.
 */
```
comment style. And it's going to be a bit difficult to reliably distinguish between this case and yours, I'm afraid...

**Comment by geographika** [comment]
Thanks @vadz - please close my pull request and use yours. 

Makes sense with the `/*` - I originally was using the SWIG docstring features with RST syntax (such as `**bold**`), and then switched to Doxygen as it was less intrusive on the C code. I found a way to automatically add immutable to the docs in the end anyway by checking for a property setter function in the Python module, so can hopefully avoid the `\*` hack.

**Comment by wsfulton** [comment]
Fixed by #1952.
